File: src/core/graph.ts
import type { SessionEvent, Graph, SessionStats } from "./types"
import { parseEventToNodes } from "./parser"

function computeStats(events: SessionEvent[]): SessionStats {
  let totalInputTokens = 0
  let totalOutputTokens = 0
  let totalCacheRead = 0
  let totalCacheCreation = 0
  let model: string | null = null

  for (const event of events) {
    const usage = event.message?.usage
    if (!usage) continue
    totalInputTokens += usage.input_tokens ?? 0
    totalOutputTokens += usage.output_tokens ?? 0
    totalCacheRead += usage.cache_read_input_tokens ?? 0
    totalCacheCreation += usage.cache_creation_input_tokens ?? 0
    if (event.message?.model) model = event.message.model
  }

  return { totalInputTokens, totalOutputTokens, totalCacheRead, totalCacheCreation, model }
}

export function buildGraph(events: SessionEvent[]): Graph {
  const nodes = events.flatMap(parseEventToNodes)
  const edges = nodes
    .filter(n => n.parentId)
    .map(n => ({
      from: n.parentId!,
      to: n.id,
      isBranch: n.branchLevel > 0,
    }))

  nodes.sort((a, b) => a.timestamp - b.timestamp)
  const stats = computeStats(events)

  return { nodes, edges, stats }
}

File: rust/graph.rs
use crate::parser::parse_event_to_node;
use crate::types::{Graph, SessionEvent};
use anyhow::Result;

pub struct GraphBuilder {
    graph: Graph,
}

impl GraphBuilder {
    pub fn new() -> Self {
        Self {
            graph: Graph::default(),
        }
    }

    pub fn build_from_events(&mut self, events: Vec<SessionEvent>) -> Result<&Graph> {
        for event in events {
            let nodes = parse_event_to_node(event)?;
            for node in nodes {
                self.graph.add_node(node);
            }
        }

        self.graph.sort_by_time();
        Ok(&self.graph)
    }

    pub fn graph(&self) -> &Graph {
        &self.graph
    }
}
